stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,282 @@
+// additional metrics.
+func (m *Metrics) Add(metrics ...MetricName) *Metrics {
+	newMetrics := make(map[MetricName]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		newMetrics[metric] = true
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// Remove removes the metrics from the metrics set and returns a new copy with
+// the metrics removed.
+func (m *Metrics) Remove(metrics ...MetricName) *Metrics {
+	newMetrics := make(map[MetricName]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		delete(newMetrics, metric)
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// MetricsOptions are the metrics options for OpenTelemetry instrumentation.
+type MetricsOptions struct {
+	// MeterProvider is the MeterProvider instance that will be used for access
+	// to Named Meter instances to instrument an application. To enable metrics
+	// collection, set a meter provider. If unset, no metrics will be recorded.
+	// Any implementation knobs (i.e. views, bounds) set in the passed in object
+	// take precedence over the API calls from the interface in this component
+	// (i.e. it will create default views for unset views).
+	MeterProvider metric.MeterProvider
+	// Metrics are the metrics to instrument. Will turn on the corresponding
+	// metric supported by the client and server instrumentation components if
+	// applicable.
+	Metrics Metrics
+	// TargetAttributeFilter is a callback that takes the target string of the
+	// channel and returns a bool representing whether to use target as a label
+	// value or use the string "other". If unset, will use the target string as
+	// is.
+	TargetAttributeFilter func(string) bool
+
+	// MethodAttributeFilter is to record the method name of RPCs handled by
+	// grpc.UnknownServiceHandler, but take care to limit the values allowed, as
+	// allowing too many will increase cardinality and could cause severe memory
+	// or performance problems. Only applicable on server side. On Client Side,
+	// pass a grpc.StaticMethodCallOption as a call option into Invoke or
+	// NewStream.
+	MethodAttributeFilter func(string) bool
+}
+
+// DialOption returns a dial option which enables OpenTelemetry instrumentation
+// code for a grpc.ClientConn.
+//
+// Client applications interested in instrumenting their grpc.ClientConn should
+// pass the dial option returned from this function as a dial option to
+// grpc.NewClient().
+//
+// For the metrics supported by this instrumentation code, a user needs to
+// specify the client metrics to record in metrics options. A user also needs to
+// provide an implementation of a MeterProvider. If the passed in Meter Provider
+// does not have the view configured for an individual metric turned on, the API
+// call in this component will create a default view for that metric.
+func DialOption(mo MetricsOptions) grpc.DialOption {
```

Comment:
This will be used for non-metrics stuff soon, right?  It should be made with that in mind.

Reply:
Yes, for tracing as well similar to OpenCensus. I think another discussion on this PR talks about tracing in the future.

Reply:
The parameter is called `MetricsOptions` but it will need to carry configuration of things besides just metrics.  So please organize the parameter and its contents with that in mind.

Reply:
Switched to nested as discussed offline.